lib/FrontendTool/FrontendTool.cpp

```diff
@@ -237,12 +237,151 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+    auto isSwiftinterface =
+      moduleFileType == file_types::TY_SwiftModuleInterfaceFile;
+
+    if (isSwiftmodule || isSwiftinterface) {
+      auto dep = pathToModuleDecl.find(depPath);
+      if (dep != pathToModuleDecl.end()) {
+        // Great, we recognize the path! Check if the file is still around.
+
+        ModuleDecl *depMod = dep->second;
+        if(depMod->isResilient() && !isSwiftinterface) {
+          SmallString<256> moduleAdjacentInterfacePath(depPath);
+          computeAdjacentInterfacePath(moduleAdjacentInterfacePath);
+          if (fs::real_path(moduleAdjacentInterfacePath, buffer, true)) {
+            err << "The module " << depMod->getName().str() << " has library"
+                << " evolution enabled but we're recording a non-adjacent"
+                << " swiftmodule at\n" << depPath << "\nin the trace.";
+            llvm::report_fatal_error(err.str());
+          }
+        }
+
+        // FIXME: Better error handling
+        StringRef realDepPath
+          = fs::real_path(depPath, buffer, /*expand_tile*/true)
+          ? StringRef(depPath) // Couldn't find the canonical path, assume
+                               // this is good enough.
+          : buffer.str();
+
+        traceInfo.push_back({
+          /*Name=*/
+          depMod->getName(),
+          /*Path=*/
+          realDepPath,
+          // TODO: There is an edge case which is not handled here.
+          // When we build a framework using -import-underlying-module, or an
+          // app/test using -import-objc-header, we should look at the direct
+          // imports of the bridging modules, and mark those as our direct
+          // imports.
+          /*IsImportedDirectly=*/
+          importedModules.find(depMod) != importedModules.end(),
+          /*SupportsLibraryEvolution=*/
+          depMod->isResilient()
+        });
+
+        continue;
+      }
+
+      // If the depTracker had an interface, that means that we must've
+      // built a swiftmodule from that interface, so we should have that
+      // filename available.
+      if (isSwiftinterface) {
+        err << "Unexpected path for swiftinterface file:\n" << depPath << "\n";
+        errorUnexpectedPath();
+      }
+
+      // Skip cached modules in the prebuiltCache, we will add the corresponding
+      // swiftinterface from the SDK directly, but this isn't checked. :-/
+      if (StringRef(depPath).contains(prebuiltCachePath))
+        continue;
+
+      // If we have a swiftmodule next to an interface, that interface path will
+      // be saved (not checked), so don't save the path to this swiftmodule.
+      SmallString<256> moduleAdjacentInterfacePath(depPath);
+      computeAdjacentInterfacePath(moduleAdjacentInterfacePath);
+      if (pathToModuleDecl.find(moduleAdjacentInterfacePath)
+          != pathToModuleDecl.end())
+        continue;
+
+      if (!fs::real_path(moduleAdjacentInterfacePath, buffer,
```

Comment:
The documentation of real_path doesn't really say that it clears buffer.. perhaps use a different local variable here or explicitly clear it?

Reply:
In a way, it doesn't really matter, because we aren't reading the buffer back. But yeah, if we were to use it, say for debugging or something, then having the code correct would be good.

Reply:
I take that back. We are reusing the buffer across loop iterations, so we should clear it after use.